### PR TITLE
#6812 ChartJS: Upgrade to V3

### LINKED ISSUE #6812: ChartJS: Upgrade to V3
Something we should keep in mind for PrimeFaces 11.
They are currently at https://github.com/chartjs/Chart.js/releases/tag/v3.0.0-beta.8.

https://www.chartjs.org/docs/master/getting-started/v3-migration

### PR DESCRIPTION
**Pull request by christophs78**
Showcase-PR: https://github.com/primefaces/primefaces-showcase/pull/335

### CODE REVIEW

**Comment by christophs78** [comment]
At the current point this PR show´s most of our showcase-examples are up and working again.
But need some more time. And not sure whether we should keep our models almost unchanged and do - where necessary - some magic in the background to map it to ChartJS 3. Maybe we should do some breaking changes on our side to better reflect the API of ChartJS 3.

**Comment by melloware** [comment]
I was afraid of that. Maybe we could mark some deprecated and do some magic behind the scenes as well?

**Comment by melloware** [comment]
RC2 was just released: https://github.com/chartjs/Chart.js/releases/tag/v3.0.0-rc.2

**Comment by christophs78** [comment]
All of our showcase-examples are up and working again.
I should take some more time to think about our API do some code-cleanup.

**Comment by melloware** [comment]
Charts 3.0.0 is final!

https://github.com/chartjs/Chart.js/releases/tag/v3.0.0

**Comment by melloware** [comment]
Already a 3.0.1: https://github.com/chartjs/Chart.js/releases/tag/v3.0.1

**Comment by melloware** [comment]
3.0.2 https://github.com/chartjs/Chart.js/releases/tag/v3.0.2

**Comment by melloware** [comment]
3.1.0: https://github.com/chartjs/Chart.js/releases/tag/v3.1.0

**Comment by melloware** [comment]
3.1.1 out: https://github.com/chartjs/Chart.js/releases/tag/v3.1.1

**Comment by christophs78** [comment]
Getting the extender-mechanism back 2 work seems to be complicated....
Got it kind of working with using window.setTimeout - trick, but far away from ideal. Some better ideas? (Probably needs at least another few hours to play around.)

```javascript
function chartExtender() {
	//copy the config options into a variable
	var options = $.extend(true, {}, this.cfg.config);

	console.log('passe ChartJS-settings an');

	options = {
		plugins: {
			"title": {"display": true, "text": "Bar Chart"}
		}
	};

	//merge all options into the main chart options
	var tmp = $.extend(true, this.cfg.config, options);
	console.log(tmp);

	// this.chart.update();
	var that = this;
	window.setTimeout(() => { updateChart(that.widgetVar); }, 10);
};

function updateChart(widgetVar) {
	let chart = PF(widgetVar).chart;
	chart.options.plugins.title.text = 'new title';
	chart.update();
}
```

**Comment by melloware** [comment]
Hmmm maybe i am confusing something but the Extender just manipulates the CFG object before the Chart is created?  Its just supposed to be a way for people to add their own code and plugins before passing it to the chart before its created `this.chart = new Chart(this.ctx, this.cfg.config); `.

Or is something different in the new Charts?

**Comment by christophs78** [comment]
`_render` is called before `init`. And probably some internal difference in ChartJS 3 which makes it a bit more complicated to change configuration after ChartJS was initialized. I´ll try to move in some way the extender from to `init` to  `_render`.

**Comment by christophs78** [comment]
> Hmmm maybe i am confusing something but the Extender just manipulates the CFG object before the Chart is created? Its just supposed to be a way for people to add their own code and plugins before passing it to the chart before its created `this.chart = new Chart(this.ctx, this.cfg.config); `.
> 
> Or is something different in the new Charts?

It was my fault. I just was temporarily confused by ChartJS 3 - data structures and had to adapt the extender - example.

**Comment by christophs78** [comment]
From my point of view this PR is read for review.

One thing someone should do some time later is to cross-check our XyzChartDataSet - classes against current ChartJS-documentation. IMO there a some gaps introduced over years during upgrading ChartJS.

**Comment by melloware** [comment]
Excellent work @christophs78

File: src/main/java/org/primefaces/component/charts/ChartRenderer.java [hunk 1]
```diff
@@ -163,70 +161,102 @@ protected void encodeScales(FacesContext context, String chartName, Object scale
             }
             hasComma = false;
 
+            boolean hasAttrs = false;
+
+            // TODO: requires some refactorings for hasAttrs
```
Comment (melloware) [inline src/main/java/org/primefaces/component/charts/ChartRenderer.java]: Remove TODO?

File: src/main/java/org/primefaces/component/charts/ChartRenderer.java [hunk 2]
```diff
@@ -163,70 +161,102 @@ protected void encodeScales(FacesContext context, String chartName, Object scale
...
+
             if (scales instanceof CartesianScales) {
                 writer.write("\"scales\":{");
                 CartesianScales cScales = (CartesianScales) scales;
-                encodeScaleCommon(writer, cScales);
+//                encodeScaleCommon(writer, cScales);
```
Comment (melloware) [inline src/main/java/org/primefaces/component/charts/ChartRenderer.java]: Delete commented line?

File: src/main/java/org/primefaces/component/charts/ChartRenderer.java [hunk 3]
```diff
@@ -163,70 +161,102 @@ protected void encodeScales(FacesContext context, String chartName, Object scale
...
+                    hasAttrs = true;
                 }
 
                 writer.write("}");
             }
             else if (scales instanceof RadialScales) {
                 writer.write("\"scale\":{");
                 RadialScales rScales = (RadialScales) scales;
-                encodeScaleCommon(writer, rScales);
+//                encodeScaleCommon(writer, rScales);
```
Comment (melloware) [inline src/main/java/org/primefaces/component/charts/ChartRenderer.java]: Delete commented line?

File: src/main/java/org/primefaces/component/charts/ChartRenderer.java [hunk 4]
```diff
@@ -163,70 +161,102 @@ protected void encodeScales(FacesContext context, String chartName, Object scale
...
     }
 
-    protected void encodeScaleCommon(ResponseWriter writer, AxesScale scale) throws IOException {
-        ChartUtils.writeDataValue(writer, "display", scale.isDisplay(), false);
-        ChartUtils.writeDataValue(writer, "weight", scale.getWeight(), true);
-    }
+//    protected void encodeScaleCommon(ResponseWriter writer, AxesScale scale) throws IOException {
+//        ChartUtils.writeDataValue(writer, "display", scale.isDisplay(), false);
+//        ChartUtils.writeDataValue(writer, "weight", scale.getWeight(), true);
+//    }
```
Comment (melloware) [inline src/main/java/org/primefaces/component/charts/ChartRenderer.java]: Delete

**Review by melloware — CHANGES_REQUESTED** [review]

**Review by melloware — APPROVED** [review]

**Comment by melloware** [comment]
3.2.0 https://github.com/chartjs/Chart.js/releases/tag/v3.2.0

**Comment by melloware** [comment]
Looks like some conflicts to resolve now.

**Comment by tandraschko** [comment]
+1 from my side if everything is working

**Comment by melloware** [comment]
@christophs78 You think this is safe to merge or do you want to upgrade to latest ChartsJS first?

**Pull request merged**